### DISCUSSION TITLE
Missing source files in database. Error: language modes specified are incompatible

### DISCUSSION THREAD

**Discussion by vinocher** [github/codeql#15272] [Q&A] [answered]
I created a database for C++ files by running the following on Windows:
`codeql database create ..\codeql\cpp-database --language=cpp --source-root=<myroot>\src  --command='cmake --build ..\build'`

The database was successfully created, but it doesn't seem to have most of the c++ source files. src.zip also doesn't have the source files.

`log\build-tracer.log` shows that the file was processed but indicates the following error for these missing files:
```
[E 17:59:33 26172] Processed command line: <root>\codeql-bundle-win64\codeql\cpp\tools\win64\extractor.exe --trapfolder ... src\settingsholder.cpp
[E 17:59:33 26172] Not using TRAP cache
Command-line error: language modes specified are incompatible
[E 17:59:33 26172] Warning[extractor-c++]: In construct_text_message: Command-line error: language modes specified are incompatible
1 catastrophic error detected in this compilation.
```
Any ideas what could be wrong?

The build was a clean build and it succeeded.

This looks similar to https://github.com/github/codeql/issues/8755, which was marked as fixed. I'm running version 2.15.5, and hit this issue.

**Comment by jketema** [comment]
Hi @vinocher,

Could you share the whole `build-tracer.log`? Without it, it's very difficult to see what might be causing this.

**Reply by vinocher** [reply]
Thanks @jketema  -- see [build-tracer.log](https://github.com/github/codeql/files/13891064/build-tracer.log). (Some names have been changed.)

I'd appreciate any tips to workaround this.

**Reply by jketema** [reply]
Thanks. We will have a look.

> I'd appreciate any tips to workaround this.

You best option is likely to use clang-cl, or the Microsoft compiler instead of clang in an emulation mode our tooling currently doesn't understand.

**Reply by vinocher** [reply]
@jketema I think clang-cl is being used (and not clang in emulation mode). Could you point to why you think it is being used in emulation mode?
Also, could you please download the build-tracer.log for future use, so I can delete it? Thanks.

**Reply by jketema** [reply]
It uses a mix. At some points the log mentions `clang.exe` and other points it mentions `clang-cl.exe`. From a quick sample it seems the former that are failing.

**Reply by vinocher** [reply]
I mainly see clang-cl.exe in the logs, whether the files are added to the database or not.

FYI, I resolved this problem by changing the compiler to msvc. Thanks for the tip. 

If anyone else faces the same issue, a workaround is to switch from clang-cl to msvc. If you're using cmake on Windows, you can do this in Power Shell by changing these variables to switch to msvc tools, or set them elsewhere:
```
$env:CMAKE_CXX_COMPILER="cl"
$env:CMAKE_C_COMPILER="cl"
$env:CMAKE_LINKER="link"
```

**Reply by jketema** [reply]
Hi @vinocher,

I understand what is going on, and believe I have a fix. However, I has some problems constructing an actual test case. May I ask where you go your clang-cl from? It seems to run in integrated cc1 mode under some circumstances (but not all), while no off-the-shelf copy of clang-cl I tried does that.

**Reply by jketema** [reply, marked as answer]
We merged a fix. This should be part of CodeQL 2.16.1, which should be released in a few weeks.

**Comment by vinocher** [comment]
> I understand what is going on, and believe I have a fix. However, I has some problems constructing an actual test case. May I ask where you go your clang-cl from? It seems to run in integrated cc1 mode under some circumstances (but not all), while no off-the-shelf copy of clang-cl I tried does that.

We get it from conda (see https://docs.conda.io/projects/miniconda/en/latest/) by installing the following packages:
  - clang=14.0.6
  - clang-tools=14.0.6
  - clangxx=14.0.6
For example, see:
https://anaconda.org/conda-forge/clang/files?version=14.0.6
https://anaconda.org/conda-forge/clang-tools/files?version=14.0.6
https://anaconda.org/conda-forge/clangxx/files?version=14.0.6

Thanks for the fix!
